spec/models/historical_event_spec.rb

```diff
@@ -5,4 +5,75 @@
     it { should belong_to(:activity) }
     it { should belong_to(:user) }
   end
+
+  describe "flexible 'value' fields which handle a range of data types" do
+    let(:event) { HistoricalEvent.new }
+
+    context "when value is a String" do
+      let(:persisted_event) do
+        set_value_fields_to("String")
+      end
+
+      it "should read back a String" do
+        expect(persisted_event.new_value).to eq("String")
+        expect(persisted_event.previous_value).to eq("String")
+      end
+    end
+
+    context "when value is an Integer" do
+      let(:persisted_event) do
+        set_value_fields_to(101)
+      end
+
+      it "should read back an Integer" do
+        expect(persisted_event.new_value).to eq(101)
+        expect(persisted_event.previous_value).to eq(101)
+      end
+    end
+
+    context "when value is a boolean" do
+      let(:persisted_event) do
+        set_value_fields_to(true)
+      end
+
+      it "should read back an boolean" do
+        expect(persisted_event.new_value).to be true
+        expect(persisted_event.previous_value).to be true
+      end
+    end
+
+    context "when value is a datetime" do
+      let(:datetime) { Time.current }
+
+      let(:persisted_event) do
+        set_value_fields_to(datetime)
+      end
+
+      it "should read back a datetime" do
+        expect(persisted_event.new_value).to eq(datetime)
+        expect(persisted_event.previous_value).to eq(datetime)
+      end
+    end
+
+    context "when value is a date" do
+      let(:date) { Date.parse("01-Jan-2020") }
+
+      let(:persisted_event) do
+        set_value_fields_to(date)
+      end
+
+      it "should read back a date" do
+        expect(persisted_event.new_value).to eq(date)
+        expect(persisted_event.previous_value).to eq(date)
+      end
+    end
+
+    def set_value_fields_to(value)
+      event.tap do |e|
+        e.new_value = value
+        e.previous_value = value
+        e.save
```

Comment:
Should we call `reload` after save here, to make sure we've fetched the object from the database again?

Reply:
hmm, interesting question! 

On revisiting I see that actually we don't need `e.save` at all. The spec passes without that persistence method call.

The serialisation behaviour we're describing is achieved through assigning and reading back (not sure exactly where!) And if we remove:

```
  serialize :new_value
  serialize :previous_value
```

from `HistoricalActivity`, that's what causes the specs to fail. Persisting or not persisting isn't relevant. 

So the example is clearer without `e.save` (or any persistence) I think?

Reply:
Out of interest, do the tests pass if we do `e.save` and `e.reload`? If so, I'm happy to do it without `e.save`

Reply:
Yes, if we save the record and reload the serialisation is still in place.

But in order to do `save` and `reload` we'd need to make the subject valid:

```
    let(:event) do
      HistoricalEvent.new(
        user: create(:delivery_partner_user), 
        activity: create(:third_party_project_activity)
      ) 
    end
```

which is more noise in addition to the unnecessary db connections.

We do have an integration test `spec/features/staff/users_can_edit_an_activity_spec.rb` which demonstrates that the units all join together as expected, i.e. we get the expected `HistoryEvent` after editing an Activity using the wizard...

Reply:
Ah, cool, course. That's fine in that case 👍